v2/cmd/controller/app/flags.go

```diff
@@ -8,77 +8,61 @@ package app
 import (
 	"flag"
 	"fmt"
-
-	"k8s.io/klog/v2"
-
-	"github.com/Azure/azure-service-operator/v2/internal/version"
 )
 
 type Flags struct {
-	MetricsAddr          string
-	ProfilingMetrics     bool
-	SecureMetrics        bool
-	HealthAddr           string
-	WebhookPort          int
-	WebhookCertDir       string
-	EnableLeaderElection bool
-	CRDManagementMode    string
-	CRDPatterns          string // This is a ';' delimited string containing a collection of patterns
+	MetricsAddr          *string
```

Comment:
It's to do with the change from `ParseFlags` to `InitFlags`, below. Previously, we set up the flagset and did the parsing all in one go, but now we need to divide the process into two parts, allowing for other flags to be added before parsing is done.

The various methods like `flagSet.String()` and `flagSet.Bool()` return a pointer to a variable that will be updated when the parsing is done. So the `*string` is not an "optional string", it is a "reference to a string".

Plus, they'll always be populated, so there are no guard clauses needed.

All that said, if there's a better solution I'm all keen because I agree this seems a bit hacky (and not in a good way).

Reply:
Doesn't `flagSet` offer the exact same API surface as `flagset` did?

Can we not just do this?

```
	var metricsAddr string

	// default here for 'MetricsAddr' is set to "0", which sets metrics to be disabled if 'metrics-addr' flag is omitted.
	flagSet.StringVar(&metricsAddr, "metrics-addr", "0", "The address the metric endpoint binds to.")
```

I tried that out locally and it seems to at least compile.

Reply:
In fact this is what we were doing before already - I guess I am not seeing why it no longer works?

Reply:
Using a local like that only works if the call to `flagSet.Parse()` happens in the same method, which it used to but no longer does.

Reply:
The variable isn't updated by the call to `flagSet.StringVar()` - it keeps the pointer, and updates the variable later, when `Parse` is called.

Reply:
Ahh I had totally missed that.

What about this approach?

```
func InitFlags(flagSet *flag.FlagSet) *Flags {
	result := &Flags{}
	
	// default here for 'MetricsAddr' is set to "0", which sets metrics to be disabled if 'metrics-addr' flag is omitted.
	flagSet.StringVar(&result.MetricsAddr, "metrics-addr", "0", "The address the metric endpoint binds to.")
	//...

	return result
}
```

Basically, rather than make the ptr-ness at each entry in the struct, we could make the whole Flags result a ptr and then use StringVar. It I think maintains the same behavior that you have where the flags aren't actually populated until parse is called, but it avoids needing N pointers and then we could either change where we pass flags to be `*Flags`, _or_ (once parse is called) we could just do something like:
```
flagsPtr := app.InitFlags(...)
...
flagSet.Parse()
...

flgs := *flagsPtr
```

and the rest of the code could be unchanged.

Reply:
This is nicer.